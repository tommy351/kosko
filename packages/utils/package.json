{
  "name": "@kosko/utils",
  "version": "0.0.0",
  "description": "Common utilities used in Kosko.",
  "homepage": "https://kosko.dev",
  "repository": "https://github.com/tommy351/kosko/tree/master/packages/yaml",
  "author": "Tommy Chen <tommy351@gmail.com>",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    "import": "./dist/index.mjs",
    "require": "./dist/index.js"
  },
  "files": [
    "/dist/"
  ],
  "keywords": [
    "kosko"
  ],
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">=14.18.0"
  },
  "dependencies": {
    "tslib": "^2.4.0"
  },
  "devDependencies": {
    "@kosko/jest-preset": "workspace:^"
  },
  "jest": {
    "preset": "@kosko/jest-preset"
  }
}
