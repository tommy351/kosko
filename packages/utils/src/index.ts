export function isRecord(
  value: unknown
): value is Record<string | symbol, unknown> {
  return value != null && typeof value === "object" && !Array.isArray(value);
}

export function toArray<T>(input: T | T[]): T[] {
  return Array.isArray(input) ? input : [input];
}

export function getErrorCode(err: unknown): string | undefined {
  if (isRecord(err) && typeof err.code === "string") {
    return err.code;
  }
}
