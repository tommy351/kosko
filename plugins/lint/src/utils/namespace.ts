import { isRecord } from "@kosko/common-utils";
import type { INamespace } from "kubernetes-models/v1/Namespace";

export function isNamespace(value: unknown): value is INamespace {
  return (
    isRecord(value) && value.apiVersion === "v1" && value.kind === "Namespace"
  );
}
