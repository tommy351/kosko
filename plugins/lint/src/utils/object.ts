import { isRecord } from "@kosko/common-utils";

export function getObjectPath<T>(
  value: unknown,
  path: readonly string[]
): T | undefined {
  let current: unknown = value;

  for (const key of path) {
    if (!isRecord(current)) return;
    current = current[key];
  }

  return current as T;
}
