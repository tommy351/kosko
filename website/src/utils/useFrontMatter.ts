import { useDoc } from "@docusaurus/plugin-content-docs/client";
import type { DocFrontMatter } from "@docusaurus/plugin-content-docs";

interface FrontMatter extends DocFrontMatter {
  available_since?: Record<string, string>;
}

export default function useFrontMatter() {
  const doc = useDoc();
  return doc.frontMatter as FrontMatter;
}
