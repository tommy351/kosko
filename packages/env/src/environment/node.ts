import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory.
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(reducers: ReducerList): ReducerExecutor;
  requireModule(environment: Environment, id: string): unknown;
  mergeValues(data: unknown[]): unknown;
}

export function createNodeEnvironment({
  cwd = process.cwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: (cb) => reducers.setReducers(cb),
    resetReducers: () => reducers.resetReducers(),
    reduce
  });

  return environment;
}
