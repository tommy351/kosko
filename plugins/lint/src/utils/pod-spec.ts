import { isRecord } from "@kosko/common-utils";
import type { IPodSpec } from "kubernetes-models/v1/PodSpec";
import { getObjectPath } from "./object";

interface PodSpecKind {
  apiVersion: string;
  kind: string;
  path: readonly string[];
}

// The following link lists all resources that contain PodTemplateSpec.
// https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podtemplatespec-v1-core
const podSpecKinds: readonly PodSpecKind[] = [
  {
    apiVersion: "v1",
    kind: "Pod",
    path: ["spec"]
  },
  {
    apiVersion: "v1",
    kind: "ReplicationController",
    path: ["spec", "template", "spec"]
  },
  {
    apiVersion: "apps/v1",
    kind: "DaemonSet",
    path: ["spec", "template", "spec"]
  },
  {
    apiVersion: "apps/v1",
    kind: "Deployment",
    path: ["spec", "template", "spec"]
  },
  {
    apiVersion: "apps/v1",
    kind: "ReplicaSet",
    path: ["spec", "template", "spec"]
  },
  {
    apiVersion: "apps/v1",
    kind: "StatefulSet",
    path: ["spec", "template", "spec"]
  },
  {
    apiVersion: "batch/v1",
    kind: "Job",
    path: ["spec", "template", "spec"]
  },
  {
    apiVersion: "batch/v1",
    kind: "CronJob",
    path: ["spec", "jobTemplate", "spec", "template", "spec"]
  }
];

interface PodSpecResult {
  kind: PodSpecKind;
  spec: IPodSpec;
}

export function getPodSpec(value: unknown): PodSpecResult | undefined {
  if (!isRecord(value)) return;

  for (const kind of podSpecKinds) {
    if (value.apiVersion !== kind.apiVersion || value.kind !== kind.kind) {
      continue;
    }

    const podSpec = getObjectPath<IPodSpec>(value, kind.path);

    if (podSpec == null) return;

    return { kind, spec: podSpec };
  }
}
