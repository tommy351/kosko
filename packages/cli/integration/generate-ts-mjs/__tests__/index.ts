import { dirname } from "node:path";
import { runNodeCLI } from "../../utils/run";

const testDir = dirname(__dirname);

test("should load loaders specified in args", async () => {
  const result = await runNodeCLI(["generate", "--loader", "ts-node/esm"], {
    cwd: testDir
  });
  expect(result.stdout).toMatchSnapshot();
});

test("should load loaders specified in config", async () => {
  const result = await runNodeCLI(
    ["generate", "--config", "kosko-loader.toml"],
    {
      cwd: testDir
    }
  );
  expect(result.stdout).toMatchSnapshot();
});

test("should load ES modules specified in args", async () => {
  const result = await runNodeCLI(
    [
      "generate",
      "--import",
      `data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));`
    ],
    { cwd: testDir }
  );
  expect(result.stdout).toMatchSnapshot();
});

test("should load ES modules specified in config", async () => {
  const result = await runNodeCLI(
    ["generate", "--config", "kosko-import.toml"],
    { cwd: testDir }
  );
  expect(result.stdout).toMatchSnapshot();
});
