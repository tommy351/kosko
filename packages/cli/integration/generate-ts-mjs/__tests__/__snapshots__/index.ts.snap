// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should load loaders specified in args 1`] = `
"apiVersion: v1
kind: Pod
metadata:
  name: test-pod"
`;

exports[`should load loaders specified in config 1`] = `
"apiVersion: v1
kind: Pod
metadata:
  name: test-pod"
`;

exports[`should use import specified in args 1`] = `
"apiVersion: v1
kind: Pod
metadata:
  name: test-pod"
`;

exports[`should use import specified in config 1`] = `
"apiVersion: v1
kind: Pod
metadata:
  name: test-pod"
`;
